### ISSUE TITLE
change activation function in Inference API for Image Classification

### ISSUE THREAD

**Issue by lombardata** [huggingface/huggingface.js#429]
Hi all,
I have trained a multilabel Image Classification model here :
[https://huggingface.co/lombardata/dino-base-2023_10_31-demo-v5
](url)
and set the task in the model card to **image classification**. This allows the automatic creation of an Inference API on the model page, but the default activation function is a Softmax.
I would like to change the activation function to a Sigmoid one (the standard activation function for Multilabel Image Classification problems), but if I try to change the `pipeline_tag` in the` Model Card`, there is no option for Multilabel Image Classification in the Computer Vision task.
Do you have any suggestions on how to do it?
Thank you in advance!

**Comment by julien-c** [comment]
(edited your post to link to the model page, @lombardata)

**Comment by julien-c** [comment]
maybe cc @osanseviero?

**Comment by osanseviero** [comment]
When a repo is tagged as `transformers` and `image-classification`, the API uses the `ImageClassificationPipeline` in `transformers` under-the-hood. It seems the pipeline always enforces softmax (https://github.com/huggingface/transformers/blob/v4.36.1/src/transformers/pipelines/image_classification.py#L122C42-L122C49). 

Let me cc @amyeroberts and @LysandreJik here. Maybe there are some changes that can be done to the `pipeline`, but it requires changes in `transformers`

**Comment by amyeroberts** [comment]
@Narsil For the pipelines, it it possible to have different tasks which map to a specific pipeline configuration. AFAICT, at the moment we have 1:1 task to settings [mapping here](https://github.com/huggingface/transformers/blob/7226f3d2b06316a1623b7b557cc32f360a854860/src/transformers/pipelines/__init__.py#L279C1-L279C1).

Could we add e.g. `"multilabel-image-classification" which uses the ImageClassificationPipeline but initializes it to use a different loss function?

**Comment by Narsil** [comment]
I don't think we should change the 1-1 assumption. It makes a lot of things much easier to reason about.

For the choice of loss we have something like that in `text-classification`: https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/text_classification.py#L207
which might be readapted. We didn't include it by default because have different loss functions based on input was confusing, but that maybe should something owned by a config.json ?

**Comment by amyeroberts** [comment]
@Narsil Makes sense! I've open #28433 to add the equivalent logic to the image classification pipeline

**Comment by lombardata** [comment]
> @Narsil Makes sense! I've open #28433 to add the equivalent logic to the image classification pipeline

Thank you all for your precious help. 
I was testing the new feature on this [model](https://huggingface.co/lombardata/dinov2-large-2024_01_05-kornia_img-size518_batch-size32_epochs70_freeze) (I have added the Image Classification tag in the model card) but still get Softmax outputs.
I have compared my config file with the PR [here](https://github.com/huggingface/transformers/pull/28433/files) and the `self.model.config.problem_type ` variable is `multi_label_classification` so that in line 171 of the PR it should select the `ClassificationFunction.SIGMOID` function, but something is going wrong.
Have a good day !

P.S. Why there is the if condition also on `self.model.config.num_labels == 1:` in the multilabel part of the PR?

**Comment by amyeroberts** [comment]
@lombardata The update was just pushed to the development branch of transformers. I'm not sure which version is used on the hub, but I'd presume the latest release. We're due to do a release this week, so should be enabled soon. 

> Why there is the if condition also on self.model.config.num_labels == 1: in the multilabel part of the PR?

In the case of binary classification, it's not necessary to have two model outputs. You can have a single model output (equivalent to config.num_labels == 1) and apply the sigmoid to that to determine if the model predicts 0 or 1.

**Comment by lombardata** [comment]
> multi_label_classification

Hi @amyeroberts , the new version of transformers was uploaded yesterday to the hub and now I could test the inference API on this [model](https://huggingface.co/lombardata/dinov2-large-2024_01_24-with_data_aug_batch-size32_epochs93_freeze) .
Unfortunately  results are different from the ones that I can get through simple inference code : 

```
# CREATE CUSTOM MODEL
def create_head(num_features , number_classes ,dropout_prob=0.5 ,activation_func =nn.ReLU):
    features_lst = [num_features , num_features//2 , num_features//4]
    layers = []
    for in_f ,out_f in zip(features_lst[:-1] , features_lst[1:]):
        layers.append(nn.Linear(in_f , out_f))
        layers.append(activation_func())
        layers.append(nn.BatchNorm1d(out_f))
        if dropout_prob !=0 : layers.append(nn.Dropout(dropout_prob))
    layers.append(nn.Linear(features_lst[-1] , number_classes))
    return nn.Sequential(*layers)
from transformers import Dinov2Config, Dinov2Model

class NewheadDinov2ForImageClassification(Dinov2ForImageClassification):
    def __init__(self, config: Dinov2Config) -> None:
        super().__init__(config)

        self.num_labels = config.num_labels
        self.dinov2 = Dinov2Model(config)

        # Classifier head
        self.classifier = create_head(config.hidden_size * 2, config.num_labels)

# IMPORT CLASSIFICATION MODEL and MODEL CONFIG PARAMETERS
model = NewheadDinov2ForImageClassification.from_pretrained(checkpoint_name)
# send model to device
model = model.to(device)

image = Image.open(image_path)
image_processor = AutoImageProcessor.from_pretrained(checkpoint_name)
# predict
inputs = image_processor(image, return_tensors="pt")
inputs = inputs.to(device)
with torch.no_grad():
    model_outputs = model(**inputs)
outputs = model_outputs["logits"][0]
outputs = outputs.cpu().numpy()
scores = sigmoid(outputs)
#print(scores)
scores = scores > 0.5         
predicted_label = list(compress(classes_names,  scores))
```
How is the model loaded whit the API ? 
I was wondering that, since I added a classification head with respect to the encoder **facebook/dinov2-large** (and changed that head wrt Dinov2ForImageClassification), maybe the model predictions with the API are wrong since weights are uploaded on the wrong architecture.
Thank you again for your constant attention !

**Comment by amyeroberts** [comment]
@lombardata The outputs are from the [image classification pipeline](https://github.com/huggingface/transformers/blob/0290ec19c901adc0f1230ebdccad11c40af026f5/src/transformers/pipelines/image_classification.py) and presumes the model can be loaded using `AutoModelForImageClassification`.
